### PR TITLE
feat(wrangler): make resources identifier optional if x-provision flag is enabled

### PR DESCRIPTION
**Pull request by edmundhung**
Fixes [DEVX-1418](https://jira.cfdata.org/browse/DEVX-1418).

This PR is the first step toward enabling automatic resource provisioning, allowing users to skip specifying resource identifiers in the Wrangler configuration. For now, this is limited to KV, R2, and D1 to keep the scope manageable.

---

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: unreleased feature

### CODE REVIEW

File: packages/wrangler/src/config/validation.ts [hunk 1]
```diff
@@ -2595,8 +2598,11 @@ const validateR2Binding: ValidatorFn = (diagnostics, field, value) => {
 		isValid = false;
 	}
 	if (
-		!isRequiredProperty(value, "bucket_name", "string") ||
-		(value as { bucket_name: string }).bucket_name.length === 0
+		getFlag("RESOURCES_PROVISION")
+			? !isOptionalProperty(value, "bucket_id", "string") ||
+				(value.bucket_id !== undefined && value.bucket_id.length === 0)
```
Comment (penalosa) [inline packages/wrangler/src/config/validation.ts:2603]: I think this should be `bucket_name`?

File: packages/wrangler/src/d1/utils.ts [hunk 2]
```diff
@@ -14,6 +14,12 @@ export function getDatabaseInfoFromConfig(
 			d1Database.database_id &&
 			(name === d1Database.database_name || name === d1Database.binding)
 		) {
+			if (!d1Database.database_name) {
+				throw new UserError(
+					`${name} bindings must have a "database_name" field`
```
Comment (penalosa) [inline packages/wrangler/src/d1/utils.ts:19]: It looks like currently they're required to have a `database_id` field only—should this error be changed?

File: packages/wrangler/src/config/validation.ts (hunk 1 above)
Comment (edmundhung) [reply packages/wrangler/src/config/validation.ts:2603]: LOL. Good catch!

Comment (edmundhung) [reply packages/wrangler/src/config/validation.ts]: https://github.com/cloudflare/workers-sdk/pull/7377/commits/73be97597160a7e01668fefe4d5a55a965ca2fad

**Review by penalosa — APPROVED** [review]

**Pull request merged**